### ISSUE TITLE
Accessing `document.cookie` in preload script will break cookies after navigating to a new URL 

### ISSUE THREAD

**Issue by davej** [electron/electron#21256]
### Preflight Checklist

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

* **Electron Version:**
  *  7.1.2
* **Operating System:**
  *  macOS 10.14.5
* **Last Known Working Electron version:**
  *  Electron 6.x

### Expected Behavior
Accessing and setting cookies should work

### Actual Behavior
If you access document.cookie in the preload script then cookies will work for the initial page load but after subsequent navigations to a new URL then cookies become broken.

### To Reproduce

1. Run this [Electron Fiddle Gist](https://gist.github.com/davej/36f3a0257c27dee1b9b5eea99ac1f86b)
2. After initial page load open the devtools
3. In the devtools console type: `navigator.cookieEnabled`
4. ✅The value returned is: `true`
5. Now navigate to any other page (for example, click the about link in the footer)
6. Again, type `navigator.cookieEnabled` in the devtools console.
7. ❌ Cookies are now broken. The value returned is: `false`

**Comment by krissss** [comment]
add `expirationDate` when use `cookies.set` #20688

**Comment by davej** [comment]
Hi @krissss, this is related do using the browser's `document.cookies`, not Electron's `cookie.set`.

I have no option to change it because `document.cookies` is being called in a third-party library.

**Comment by ronjouch** [comment]
Users of Nativefier (a helper to bundle websites as Electron apps) are affected by this: https://github.com/jiahaog/nativefier/issues/902 .

@ckerr additional data to precise the labels you added to the issue:

1. This is not mac-specific. Reproduced on Linux too.
2. I confirm the regression window is [6.1.9, 7.0.1]:
    - In 6.1.9, `navigator.cookieEnabled` is always `true`.
    - But in 7.0.1 (and later versions, issue still reproduced in latest 8.1.1), `navigator.cookieEnabled` is `true` initially, but flips to `false` on subsequent page loads, causing some sites to break.

Like @davej, I (the Electron app) am *not* setting anything; the word `cookie` is actually absent from the whole codebase. Is there something I can do as Electron user to flip this to `true` on each navigation? Can my Electron code modify this property to un-break things?

Thanks for the help.

**Comment by felixrieseberg** [comment]
Do things still break when you access `document.cookie` once the `loaded` event has fired? Doing so before shouldn't break things, but accessing cookies _before_ navigation has technically occurred feels wrong.

**Comment by deepak1556** [comment]
@felixrieseberg you are correct accessing `document.cookie` after document loaded will work fine in preload script. 

I ran the example with debug build and the following crash happened in network service process

```
[13208:0415/142206.741440:FATAL:restricted_cookie_manager.cc(516)] Check failed: top_frame_origin_ok. top_frame_origin from renderer='https://github.com' from browser='null [internally: (844A589F1B5B08112EF0A33EFF452737) anonymous]';
0   Electron Framework                  0x0000000123d3e0a9 base::debug::CollectStackTrace(void**, unsigned long) + 9
1   Electron Framework                  0x0000000123991953 base::debug::StackTrace::StackTrace() + 19
2   Electron Framework                  0x00000001239f21e5 logging::LogMessage::~LogMessage() + 837
3   Electron Framework                  0x000000012a158bcf network::RestrictedCookieManager::ValidateAccessToCookiesAt(GURL const&, net::SiteForCookies const&, url::Origin const&) + 1055
4   Electron Framework                  0x000000012a158220 network::RestrictedCookieManager::GetAllForUrl(GURL const&, net::SiteForCookies const&, url::Origin const&, mojo::InlinedStructPtr<network::mojom::CookieManagerGetOptions>, base::OnceCallback<void (std::__1::vector<net::CanonicalCookie, std::__1::allocator<net::CanonicalCookie> > const&)>) + 384
5   Electron Framework                  0x000000012a15efa3 network::RestrictedCookieManager::GetCookiesString(GURL const&, net::SiteForCookies const&, url::Origin const&, base::OnceCallback<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)>) + 1011
6   Electron Framework                  0x00000001173e367a network::mojom::RestrictedCookieManagerStubDispatch::AcceptWithResponder(network::mojom::RestrictedCookieManager*, mojo::Message*, std::__1::unique_ptr<mojo::MessageReceiverWithStatus, std::__1::default_delete<mojo::MessageReceiverWithStatus> >) + 5690
7   Electron Framework                  0x000000012a079561 network::mojom::RestrictedCookieManagerStub<mojo::UniquePtrImplRefTraits<network::mojom::RestrictedCookieManager, std::__1::default_delete<network::mojom::RestrictedCookieManager> > >::AcceptWithResponder(mojo::Message*, std::__1::unique_ptr<mojo::MessageReceiverWithStatus, std::__1::default_delete<mojo::MessageReceiverWithStatus> >) + 289
8   Electron Framework                  0x000000012558de24 mojo::InterfaceEndpointClient::HandleValidatedMessage(mojo::Message*) + 2852
9   Electron Framework                  0x000000012559b6b4 mojo::MessageDispatcher::Accept(mojo::Message*) + 724
10  Electron Framework                  0x0000000125592408 mojo::InterfaceEndpointClient::HandleIncomingMessage(mojo::Message*) + 280
11  Electron Framework                  0x00000001255ac9b9 mojo::internal::MultiplexRouter::ProcessIncomingMessage(mojo::internal::MultiplexRouter::MessageWrapper*, mojo::internal::MultiplexRouter::ClientCallBehavior, base::SequencedTaskRunner*) + 3177
12  Electron Framework                  0x00000001255aaad6 mojo::internal::MultiplexRouter::Accept(mojo::Message*) + 886
13  Electron Framework                  0x000000012559b78e mojo::MessageDispatcher::Accept(mojo::Message*) + 942
14  Electron Framework                  0x0000000125574df1 mojo::Connector::DispatchMessage(mojo::Message) + 1233
15  Electron Framework                  0x0000000125577d11 mojo::Connector::ReadAllAvailableMessages() + 1505
16  Electron Framework                  0x00000001255773c3 mojo::Connector::OnHandleReadyInternal(unsigned int) + 451
17  Electron Framework                  0x00000001148613db mojo::SimpleWatcher::DiscardReadyState(base::RepeatingCallback<void (unsigned int)> const&, unsigned int, mojo::HandleSignalsState const&) + 331
18  Electron Framework                  0x000000012561c83d mojo::SimpleWatcher::OnHandleReady(int, unsigned int, mojo::HandleSignalsState const&) + 1021
19  Electron Framework                  0x000000012561d5e8 mojo::SimpleWatcher::Context::Notify(unsigned int, MojoHandleSignalsState, unsigned int) + 1944
20  Electron Framework                  0x0000000125619d70 mojo::SimpleWatcher::Context::CallNotify(MojoTrapEvent const*) + 160
21  Electron Framework                  0x0000000119658712 mojo::core::WatcherDispatcher::InvokeWatchCallback(unsigned long, unsigned int, mojo::core::HandleSignalsState const&, unsigned int) + 546
22  Electron Framework                  0x0000000119656f74 mojo::core::Watch::InvokeCallback(unsigned int, mojo::core::HandleSignalsState const&, unsigned int) + 484
23  Electron Framework                  0x0000000119646525 mojo::core::RequestContext::~RequestContext() + 2405
24  Electron Framework                  0x000000011961775f mojo::core::NodeChannel::OnChannelMessage(void const*, unsigned long, std::__1::vector<mojo::PlatformHandle, std::__1::allocator<mojo::PlatformHandle> >) + 4735
25  Electron Framework                  0x00000001195dbc8d mojo::core::Channel::TryDispatchMessage(base::span<char const, 18446744073709551615ul>, unsigned long*) + 1917
26  Electron Framework                  0x000000011966934a mojo::core::(anonymous namespace)::ChannelMac::OnMachMessageReceived(unsigned int) + 6922
27  Electron Framework                  0x0000000123df88ad base::MessagePumpKqueue::ProcessEvents(int) + 2109
28  Electron Framework                  0x0000000123df5255 base::MessagePumpKqueue::DoInternalWork(base::MessagePump::Delegate::NextWorkInfo*) + 981
29  Electron Framework                  0x0000000123df4bb6 base::MessagePumpKqueue::Run(base::MessagePump::Delegate*) + 326
30  Electron Framework                  0x0000000123bbf3cf base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::Run(bool, base::TimeDelta) + 767
31  Electron Framework                  0x0000000123adced3 base::RunLoop::Run() + 1987
32  Electron Framework                  0x0000000123c5ecd3 base::Thread::Run(base::RunLoop*) + 451
33  Electron Framework                  0x0000000123c5f68e base::Thread::ThreadMain() + 2270
34  Electron Framework                  0x0000000123d6e89e base::(anonymous namespace)::ThreadFunc(void*) + 206
35  libsystem_pthread.dylib             0x00007fff69b23109 _pthread_start + 148
36  libsystem_pthread.dylib             0x00007fff69b1eb8b thread_start + 15
```

Essentially chrome has now stricter check for cookie access based on the origin of the document https://chromium-review.googlesource.com/c/chromium/src/+/1635430 , this was in preparation for samesite cookie project https://www.chromium.org/updates/same-site

In short if you try to access `document.cookie` before page load, the origin is essentially null and hence leads all sorts of disaster.

**Comment by nornagon** [comment]
Can we degrade in a more graceful way or is this just WONTFIX then?

**Comment by deepak1556** [comment]
@nornagon chromium already does its best to exit this bad access in a graceful way https://source.chromium.org/chromium/chromium/src/+/master:services/network/restricted_cookie_manager.cc;l=509-553 , the network process is restarted and other loads continue. Anything more would be a dangerous patch touching `network::RestrictedCookieManager`.

**Comment by nornagon** [comment]
restarting the whole network process seems like a pretty serious fail but at least it's localized.

how does chromium not encounter this with something like:

```
const child = window.open('')
child.document.cookie
```
?

https://crbug.com/983090 suggests that there were some missing renderer-side checks, e.g. https://chromium-review.googlesource.com/c/chromium/src/+/1715601 which adds one in. Are we somehow skipping those?

**Comment by deepak1556** [comment]
It would be possible if we are running the preload script right after the securityorigin is set on the renderer side https://github.com/electron/electron/blob/master/patches/chromium/add_didinstallconditionalfeatures.patch#L77 but definitely before the origin is set for the browser side.

**Comment by deepak1556** [comment]
I think the safest fix here is to avoid using any document apis until the page has been loaded, so that the origin checks are respected on both renderer and browser.

**Comment by miniak** [comment]
Electron 9 is no longer affected by this issue

**Comment by electron-triage** [comment]
The Electron version reported on this issue is no longer supported. See our [supported versions documentation](https://www.electronjs.org/docs/tutorial/support#supported-versions).

If this is still reproducible on a supported version, please open a new issue with any other new information that a maintainer should know.

Thank you for taking the time to report this issue and helping to make Electron better! Your help is appreciated.

**Comment by miniak** [comment]
interestingly this issue no longer reproduced in Electron 19+

**Comment by davej** [comment]
Can confirm, this issue seems to have been fixed.
